opp_gui/src/widgets/tool_path_parameters_editor_widget.cpp

```diff
@@ -209,6 +312,65 @@ void ToolPathParametersEditorWidget::onGenerateToolPathsComplete(
   }
 }
 
+
+void ToolPathParametersEditorWidget::onGenerateHeatToolPathsComplete(
+    const actionlib::SimpleClientGoalState& state,
+    const heat_msgs::GenerateHeatToolPathsResultConstPtr& res)
+{
+  for (int i = progress_dialog_->minimum(); i < progress_dialog_->maximum(); ++i)
+  {
+    progress_dialog_->setValue(i);
+    ros::Duration(0.01).sleep();
+  }
+  progress_dialog_->hide();
+
+  if (state != actionlib::SimpleClientGoalState::SUCCEEDED)
+  {
+    std::string message = "Action '" + GENERATE_HEAT_TOOLPATHS_ACTION + "' failed to succeed";
+    QMessageBox::warning(this, "Tool Path Planning Error", QString(message.c_str()));
```

Comment:
Creating a Qt message box in a function that is called by the ROS thread rather than the Qt thread (this function will be called by the action server when it completes) will crash your program due to "recursive painting". You can get around this by emitting a signal with your message and connecting it with a separate function in this class that creates the message dialog box. The trigger via signal makes sure that the message box is created in the Qt thread

Reply:
Please take this one off line too. I don't understand you here either.